### PR TITLE
CCS: skip empty search hits when minimizing round-trips

### LINKED ISSUE #40067: CCS: field collapsing doesn't work when minimizing roundtrips and remote index is empty
**Elasticsearch version** (`bin/elasticsearch --version`): latest master snapshot (hash: `8839a72`)

**Plugins installed**: []

**JVM version** (`java -version`): `java version "11.0.1" 2018-10-16 LTS`

**OS version** (`uname -a` if on a Unix-like system): `Darwin Kernel Version 18.2.0`

**Description of the problem including expected versus actual behavior**:

When performing a CCS search request, an error occurs if there is field collapsing in the query.

The exact error is:
```
java.lang.ArrayStoreException: arraycopy: element type mismatch: can not cast one of the elements of java.lang.Object[] to the type of the destination array, org.apache.lucene.search.grouping.CollapseTopFieldDocs
	at java.util.ArrayList.toArray(ArrayList.java:432) ~[?:?]
	at org.elasticsearch.action.search.SearchPhaseController.mergeTopDocs(SearchPhaseController.java:236) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.action.search.SearchResponseMerger.getMergedResponse(SearchResponseMerger.java:190) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.action.search.TransportSearchAction$3.createFinalResponse(TransportSearchAction.java:389) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
	at org.elasticsearch.action.search.TransportSearchAction$3.createFinalResponse(TransportSearchAction.java:379) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
```

**Steps to reproduce**:

1. Start two ES instances, ensuring they have unique `cluster.name`s
2. Configure one ES instance to talk to the other through CCS
```
POST _cluster/settings
{
  "persistent": {
    "cluster": {
      "remote": {
        "other": {
          "seeds": [
            "{other_es_instance_address}"
          ]
        }
      }
    }
  }
}
```
3. Put data in the an index on the ES cluster you ran the above query:
```
POST _bulk
{ "index": { "_index": "sales" } }
{ "price": 30, "payment_type": "credit_card" }
{ "index": { "_index": "sales" } }
{ "price": 40, "payment_type": "cash" }
```
4. **DO NOT** put any data into the other ES cluster
5. Perform a CCS search that uses field collapsing from the same ES instance
```
POST *:sales*,sales*/_search
{
  "collapse": {
    "field": "payment_type.keyword"
  }
}
```

This error also happens if you modify step 4 to *only* put the index (but do not index documents into it) on the other ES instance:
```
PUT sales
{
  "mappings": {
    "properties" : {
      "payment_type" : {
        "type" : "text",
        "fields" : {
          "keyword" : {
            "type" : "keyword",
            "ignore_above" : 256
          }
        }
      },
      "price" : {
        "type" : "long"
      }
    }
  }
}
```

The error *DOES NOT* happen if you actually index documents into the index on the other ES cluster:
```
POST _bulk
{ "index": { "_index": "sales" } }
{ "price": 30, "payment_type": "credit_card" }
{ "index": { "_index": "sales" } }
{ "price": 40, "payment_type": "cash" }

```

This is currently **BREAKING** the Kibana Stack Monitoring app, as we run these sorts of field collapsing which are currently broken with CCS enabled.

Related to #40059

### PR DESCRIPTION
**Pull request by javanna**
When minimizing round-trips in CCS, each cluster returns its own independent
search response. In case sort by field and/or field collapsing were
requested, when one cluster has no results to return, the information
about the field that sorting was based on (`SortField` array) as well as
the field (and the values) that collapsing was performed on are missing
in the search response. That causes problems as we can't build the
proper `TopDocs` instance for empty results which would need to be 
either `TopFieldDocs` or `CollapseTopFieldDocs`. The merge routine 
expects all the top docs of the same exact type which can't be guaranteed 
when some cluster returns results and some other doesn't. Given that
the problematic results are empty, hence have no impact on the final
results, we can simply skip them.

Relates to #32125
Closes #40067

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/action/search/SearchResponseMerger.java [hunk 1]
```diff
@@ -178,10 +178,17 @@ SearchResponse getMergedResponse(Clusters clusters) {
                 assert trackTotalHits == null || trackTotalHits;
                 trackTotalHits = true;
             }
-            TopDocs topDocs = searchHitsToTopDocs(searchHits, totalHits, shards);
-            topDocsStats.add(new TopDocsAndMaxScore(topDocs, searchHits.getMaxScore()),
-                searchResponse.isTimedOut(), searchResponse.isTerminatedEarly());
-            topDocsList.add(topDocs);
+            if (totalHits.value > 0) {
```
Comment (jimczi) [inline server/src/main/java/org/elasticsearch/action/search/SearchResponseMerger.java]: If track_total_hits is disabled you can have totalHits.value equals to 0 even though searchHits are not empty. I think something like this should work:
````
TopDocs topDocs = searchHitsToTopDocs(searchHits, totalHits, shards);
 topDocsStats.add(new TopDocsAndMaxScore(topDocs, searchHits.getMaxScore()),
                searchResponse.isTimedOut(), searchResponse.isTerminatedEarly());
if (searchHits.getHits().length == 0) {
  topDocsList.add(topDocs);
}
````
?

**Review by jimczi — CHANGES_REQUESTED** [review]
Thanks @javanna , I left one comment regarding the fix when track_total_hits is disabled.

File: server/src/main/java/org/elasticsearch/action/search/SearchResponseMerger.java (hunk 1 above)
Comment (javanna) [reply server/src/main/java/org/elasticsearch/action/search/SearchResponseMerger.java]: I was going to ask you about this, not sure. Let me try it.

Comment (javanna) [reply server/src/main/java/org/elasticsearch/action/search/SearchResponseMerger.java]: works, I pushed an update. I also had to adjust topDocsToSearchHits to accept null TopDocs which happens when all clusters return empty search hits.

File: server/src/test/java/org/elasticsearch/action/search/SearchResponseMergerTests.java [hunk 2]
```diff
@@ -570,6 +570,72 @@ public void testMergeNoResponsesAdded() {
...
+                ShardSearchFailure.EMPTY_ARRAY, SearchResponse.Clusters.EMPTY);
+            merger.add(searchResponse);
+        }
+        {
+            SearchHits empty = new SearchHits(new SearchHit[0], new TotalHits(0, TotalHits.Relation.EQUAL_TO), Float.NaN, null, null, null);
+            InternalSearchResponse response = new InternalSearchResponse(empty, null, null, null, false, false, 1);
+            SearchResponse searchResponse = new SearchResponse(response, null, 1, 1, 0, 1L,
+                ShardSearchFailure.EMPTY_ARRAY, SearchResponse.Clusters.EMPTY);
+            merger.add(searchResponse);
+        }
```
Comment (jimczi) [inline server/src/test/java/org/elasticsearch/action/search/SearchResponseMergerTests.java]: Maybe add a test case that sets the totalHits to `null` and a non-empty `SearchHits` to check the new logic ?

**Review by jimczi — APPROVED** [review]
LGTM, I left a small comment regarding the tests.

File: server/src/test/java/org/elasticsearch/action/search/SearchResponseMergerTests.java (hunk 2 above)
Comment (javanna) [reply server/src/test/java/org/elasticsearch/action/search/SearchResponseMergerTests.java]: that's already tested as part of the existing testMergeSearchHits. In fact that test was failing with my first buggy version ;)

Comment (jimczi) [reply server/src/test/java/org/elasticsearch/action/search/SearchResponseMergerTests.java]: ok thanks

**Pull request merged**